부산대 BE_오선재 - Step1 유효성 검사 및 예외 처리

5seonjae opened the pull request:
🔀 Step 1 - 유효성 검사 및 예외 처리

✅ 구현 내용

- 상품 등록/수정 요청 시 유효성 검사 로직 추가

    - @NotBlank, @Size, @Pattern, @Min, @NotNull 등 Bean Validation 기반

    - 상품명 최대 15자 제한

    - 허용된 특수문자 외 입력 방지

    - 이미지 URL 형식 (http://, https://) 검증

- "카카오" 문구 포함 상품명에 대한 추가 검증 기능 구현

    - DB에 사전 등록된 승인 상품만 등록 가능

    - 승인되지 않은 경우 IllegalArgumentException 발생

- 전역 예외 처리기(@ControllerAdvice) 구현

    - MethodArgumentNotValidException, IllegalArgumentException, NoSuchElementException 처리

    - JSON API 응답 및 HTML 응답에서 모두 처리

- HTML 폼 페이지에서 에러 메시지 사용자에게 출력

    - 상품 등록/수정 실패 시 입력값 유지 및 에러 메시지 표시

- 상품 등록 기능에 대한 E2E 테스트 작성

    - 정상 등록, 유효성 실패, 승인 상품 등록 여부 확인

- 추가 수정 사항

    - Repository 코드를 JdbcTemplate 기반에서 JdbcClient 로 수정했습니다.

📄 관련 문서

- 상품 관리 1단계 README: [spring-gift-product/README.md](https://github.com/5seonjae/spring-gift-product/blob/step3/README.md)

- 현재 단계 README: [step1/README.md](https://github.com/5seonjae/spring-gift-wishlist/blob/step1/README.md)

---

🙋‍♂️ 리뷰 요청 사항 및 질문

이번 Step1 미션(유효성 검사 및 예외 처리)을 진행하며 몇 가지 궁금한 점이 있어 리뷰어님께 질문드립니다:

1️⃣ E2E 테스트 작성 방식에 대해

MockMvc를 이용해 HTML 폼 전송 시나리오 기반으로 테스트를 작성했는데,
제가 구현한 방식이 E2E 테스트의 목적과 방향에 부합하는지 궁금합니다.

- 상품 등록 정상 케이스

- 유효성 실패 케이스 (빈 값, 길이 초과, 가격 음수 등)

- '카카오' 문구 포함 상품 등록 시 MD 승인 여부에 따른 분기

2️⃣ 테스트 실행 시 나타나는 경고 문구

테스트를 실행할 때 아래와 같은 Mockito 관련 경고 메시지가 출력되는데요:

Mockito is currently self-attaching to enable the inline-mock-maker...
WARNING: A Java agent has been loaded dynamically...

아직 Mockito를 직접 사용하지는 않았는데,
해당 경고가 왜 발생하는지, 그리고 어떻게 대응하면 좋을지 조언 부탁드립니다.

(참고 링크는 다음과 같습니다:
https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/org/mockito/Mockito.html#0.3)

3️⃣ '카카오' 문구 포함 상품명 검증 방식에 대해

미션 요구사항에서 '카카오' 포함 상품명은 MD 승인 필요 라는 설명만 있었고,
구체적인 검증 방식은 정해져 있지 않아 저만의 방식으로 커스터마이징하여 구현해봤습니다.

🔍 구현 방식

- MD가 관리하는 '카카오' 상품 등록/수정 관련 허용 상품 데이터 테이블 approved_products ( id, name ) 을 DB에 생성
( 만약 MD와 협의를 어떤 상품에 대해 협의를 한 상황이라면 approved_products 에 해당 상품의 이름이 들어있음 )

- 상품을 등록하거나 수정할 때, 등록하거나 수정하려는 상품의 이름에 '카카오' 라는 문구가 존재한다면 approved_products의 상품들의 이름을 조회해서 비교

- 등록/수정하려는 상품이 approved_products 테이블에 존재한다면 등록/수정이 정상적으로 가능

- 등록/수정하려는 상품이 approved_products 테이블에 존재하지 않는다면 IllegalArgumentException을 발생시켜 등록/수정 차단

- HTML 폼에서는 등록/수정 폼으로 다시 리다이렉션 한 뒤 "'카카오'가 포함된 상품은 MD 승인이 필요합니다." 메세지로 사용자에게 피드백 제공

❓ 질문

- 위 방식처럼 별도의 승인 목록 테이블을 만들고, 이름 기반으로 검증하는 방식이 괜찮은 설계인지

- 더 나은 방식이 있다면 조언을 부탁드립니다

5seonjae committed:
fix: validation 의존성 추가

5seonjae committed:
docs: 위시 리스트 Step 2 및 추가 구현 기능에 대한 README.md 작성

5seonjae committed:
chore: data.sql 초기 데이터 이미지 URL 수정

5seonjae committed:
feat: ProductViewRequestDto에 상품명 최대 길이 15자 제한 검증 기능 추가

5seonjae committed:
feat: ProductCreateRequestDto에 상품명 최대 길이 15자 제한 검증 기능 추가

5seonjae committed:
feat: ProductCreateRequestDto에 상품명 최대 길이 15자 제한 검증 기능 추가

5seonjae committed:
feat: ProductViewRequestDto의 상품명에 공백을 포함한 허용된 특수 문자 검증 기능 추가

- 허용된 특수 문자 : ( ), [ ], +, -, &, /, _

5seonjae committed:
feat: ProductUpdateRequestDto의 상품명에 공백을 포함한 허용된 특수 문자 검증 기능 추가

- 허용된 특수 문자 : ( ), [ ], +, -, &, /, _

5seonjae committed:
feat: ProductCreateRequestDto의 상품명에 공백을 포함한 허용된 특수 문자 검증 기능 추가

- 허용된 특수 문자 : ( ), [ ], +, -, &, /, _

5seonjae committed:
docs: 유효성 검사 및 예외 처리 기능 문서화 (README.md 업데이트)

5seonjae committed:
feat: '카카오' 문구가 포함된 상품명을 등록할 수 있는 승인 리스트 테이블 생성

5seonjae committed:
feat: ApprovedProduct 엔티티 클래스 생성

5seonjae committed:
feat: ApprovedProductRepository 클래스 생성

5seonjae committed:
feat: ApprovedProductRepository - 승인 여부 확인 메서드 isApproved 구현

5seonjae committed:
feat: ApprovedProductRepository - 승인 상품 등록 메서드 save 구현

5seonjae committed:
feat: ApprovedProductRepository - @Repository 애너테이션 적용

5seonjae committed:
feat: ProductService에 ApprovedProductRepository DI

5seonjae committed:
feat: '카카오' 문구 유효성 검증 메서드 추가

5seonjae committed:
feat: 상품 등록/수정 시 '카카오' 검증 로직 적용

5seonjae committed:
feat: 승인 상품 초기 데이터 등록 (data.sql)

5seonjae committed:
feat: 폼 기반 상품 등록 시 예외 메시지 화면 출력 기능 추가

- 상품명에 "카카오" 포함 시 승인되지 않은 경우 IllegalArgumentException 발생
- 예외 발생 시 사용자에게 메시지를 표시하도록 컨트롤러에서 model에 errorMessage 추가
- 유효성 검사 통과 후 비즈니스 로직 중 발생하는 예외도 사용자에게 안내 가능

5seonjae committed:
feat: 상품 등록 폼에 예외 메시지 출력 UI 추가

- errorMessage 모델 속성이 존재할 경우 상단에 경고 메시지 표시
- "카카오" 미승인 상품 등록 시 사용자에게 메시지 안내

5seonjae committed:
chore: 승인 상품명 '카카오 프렌즈 필통'으로 데이터 수정

5seonjae committed:
feat: 상품 수정 실패 시 IllegalArgumentException 메시지를 모델에 추가하여 폼에 노출

5seonjae committed:
test: ProductViewControllerTest 클래스 생성 및 테스트 환경 설정

- @SpringBootTest, @AutoConfigureMockMvc, @Transactional 적용
- MockMvc를 이용한 컨트롤러 테스트 기반 구성 완료

5seonjae committed:
feat: 테스트용 ApprovedProduct(String name) 생성자 추가

5seonjae committed:
test: 승인된 '카카오' 상품 등록 성공 테스트 추가

- ApprovedProductRepository를 활용해 사전 승인 상품 등록
- '카카오' 문구가 포함된 상품명이 성공적으로 등록되고 리다이렉트되는 흐름 검증
- 상태 코드 302 Found 및 리다이렉트 URL("/admin/products") 확인

5seonjae committed:
test: 승인되지 않은 '카카오' 상품 등록 실패 테스트 추가

- 승인 목록에 없는 '카카오' 포함 상품명을 등록할 경우 폼 화면으로 돌아가는 흐름 검증
- 상태 코드 200 OK, 뷰 이름 'products/form', 모델에 'errorMessage' 존재 여부 확인

5seonjae committed:
test: 상품명 누락 시 등록 실패 테스트 추가

- name 파라미터가 비어 있을 경우 유효성 검증 실패 처리 검증
- 상태 코드 200 OK, 'products/form' 뷰 반환, name 필드에 바인딩 에러 발생 여부 확인

5seonjae committed:
test: 상품명 15자 초과 시 등록 실패 테스트 추가

- 상품명이 16자인 경우 유효성 검증 실패 처리 확인
- 상태 코드 200 OK, 'products/form' 뷰 반환, name 필드에 바인딩 에러 발생 여부 검증

5seonjae committed:
test: 가격이 음수일 경우 등록 실패 테스트 추가

- price 값이 -500인 경우 유효성 검증 실패 처리 확인
- 상태 코드 200 OK, 'products/form' 뷰 반환, price 필드에 바인딩 에러 발생 여부 검증

5seonjae committed:
test: 빈 문자열 가격 입력 시 등록 실패 테스트 추가

- price 파라미터가 ""일 경우 Integer 변환 실패로 바인딩 오류 발생
- 상태 코드 200 OK, 'products/form' 뷰 반환, price 필드 바인딩 에러 검증

5seonjae committed:
test: 유효하지 않은 이미지 URL 입력 시 등록 실패 테스트 추가

- http/https로 시작하지 않는 imageUrl 값에 대해 @Pattern 유효성 검증 실패 확인
- 상태 코드 200 OK, 'products/form' 뷰 반환, imageUrl 필드에 바인딩 에러 발생 여부 검증

5seonjae committed:
test: 이미지 URL 누락 시 등록 실패 테스트 추가

- imageUrl 파라미터가 빈 문자열("")일 경우 @Pattern 유효성 검증 실패 확인
- 상태 코드 200 OK, 'products/form' 뷰 반환, imageUrl 필드에 바인딩 에러 발생 여부 검증

5seonjae committed:
docs: 유효성 검증 및 테스트 항목 README 체크

5seonjae committed:
docs: README 내 @Pattern 정규표현식 이스케이프 처리

5seonjae committed:
refactor: JdbcTemplate → JdbcClient로 ApprovedProductRepository 리팩토링

- JdbcTemplate 기반 쿼리 코드를 JdbcClient로 변경
- 승인 여부 확인 및 승인 상품 등록 로직을 JdbcClient 방식으로 재작성
- H2 환경에서 NamedParameter 이슈를 피하기 위해 positional parameter (?) 사용

5seonjae committed:
refactor: ProductRepository의 데이터 접근 로직을 JdbcClient 기반으로 리팩토링

- JdbcTemplate 기반 CRUD 메서드를 JdbcClient API로 전환
- ResultSet 매핑 로직은 기존 mapRowToProduct 메서드 재사용

5seonjae committed:
docs: README에 Repository 리팩토링(JdbcTemplate → JdbcClient) 내용 반영

hjjae2 commented on src/main/java/gift/controller/ProductViewController.java:
```diff
@@ -92,13 +102,22 @@ public String updateProduct(@PathVariable Long id,
+        try {
+
+            // dto를 변환해서 넘김
+            ProductUpdateRequestDto updateDto = new ProductUpdateRequestDto(
+                dto.getName(), dto.getPrice(), dto.getImageUrl()
+            );
```
`ProductUpdateRequestDto` 와 `ProductViewRequestDto` 는 어떤 차이가 있나요?

5seonjae replied:
### **ProductUpdateRequestDto 와 ProductViewRequestDto 의 차이**

프로젝트 구조와 실제 사용 위치를 살펴보면 ProductViewRequestDto(뷰 전용)와 ProductUpdateRequestDto(API 전용) 이 자연스럽게 역할이 갈립니다.

- 사용 컨트롤러

    ProductViewRequestDto : ProductViewController - 어드민 HTML 폼 바인딩에서 신규 등록·폼 기반 수정에 사용

    ProductUpdateRequestDto : ProductController 및 ProductViewController 내부의 서비스 호출부 - JSON Create/Update API, 또는 뷰 폼에서 변환해 서비스로 전달할 때 사용

- 바인딩 방식

    ProductViewRequestDto : 기본 생성자 + Setter ➜ Spring MVC @ModelAttribute 폼 바인딩

    ProductUpdateRequestDto : 모든 필드를 생성자 주입 (Setter 없음) ➜ 서비스 계층부터는 불변 객체로 다루기 위함

- 식별자 필드

    ProductViewRequestDto : 없음 (신규 등록·폼 수정 시 클라이언트가 ID를 직접 주지 않음)

    ProductUpdateRequestDto : 컨트롤러 경로 변수로 받는 id와 세트로 사용 (DTO 안에는 식별자 필드가 없음)

- 확장 범위

    ProductViewRequestDto : 썸네일 MultipartFile 등 뷰 전용 필드가 늘어날 가능성

    ProductUpdateRequestDto : 부분 업데이트 (PATCH) 플래그, 승인 상태 등 API 전용 필드가 추가될 가능성

즉, 계층별 책임(뷰 ↔ API)과 객체 불변성(Setter 여부) 요구사항이 달라서 두 클래스를 분리했습니다.

hjjae2 commented on src/main/java/gift/repository/ApprovedProductRepository.java:
```diff
@@ -0,0 +1,33 @@
+                .param(name)
+                .query(Integer.class)
+                .single();
+        return count != null && count > 0;
```
`count` 는 null 이 될 가능성이 없다고 하네요, `single()` 메서드의 설명을 한번 살펴보시겠어요? 😄

5seonjae replied:
당시엔 깊이 고민하지 않고 single() 을 썼는데, 리뷰 덕분에 메서드 설명을 다시 확인했습니다.

- single() 은 결과가 0 row면 EmptyResultDataAccessException, 2 row 이상이면 IncorrectResultSizeDataAccessException을 던지기 때문에 null 반환이 불가능합니다.
- COUNT(*)는 테이블이 비어 있어도 0을 담은 1 row를 돌려주므로 DB 측에서도 null 이 나올 수 없습니다.

```java
Integer count = jdbcClient
        .sql("SELECT COUNT(*) FROM approved_products WHERE name = ?")
        .param(name)
        .query(Integer.class)
        .single();

return count > 0;   // null 체크 제거 예정
```

추가로 승인 여부 확인 메서드(isApproved)에 대해 개선할 수 있는 점들을 생각해봤습니다.

- 스칼라 값만 필요하니 singleValue() + 원시형(int.class)을 쓰면 null 체크 없이 의도가 더 명확해집니다.
- 성능과 가독성을 고려해 SELECT EXISTS(…) + query(Boolean.class).single()로 바꾸는 방법도 검토하겠습니다.
    (DB가 첫 행 발견 즉시 종료하므로 COUNT(*) 보다 빠릅니다.)

hjjae2 replied:
훌륭하시네요. 👍

hjjae2 replied:
> 스칼라 값만 필요하니 singleValue() + 원시형(int.class)을 쓰면 null 체크 없이 의도가 더 명확해집니다.
> 성능과 가독성을 고려해 SELECT EXISTS(…) + query(Boolean.class).single()로 바꾸는 방법도 검토하겠습니다.
> (DB가 첫 행 발견 즉시 종료하므로 COUNT(*) 보다 빠릅니다.)

검토해주신 부분들도 여유가 되면 적용해봐주시면 좋겠어요. 👍

hjjae2 commented on src/main/java/gift/repository/ProductRepository.java:
```diff
@@ -1,30 +1,31 @@
 public class ProductRepository {
 
-    private final JdbcTemplate jdbcTemplate;
+    private final JdbcClient jdbcClient;
```
`JdbcClient` 를 써보시는군요. 좋네요.👍

hjjae2 commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -13,12 +14,16 @@
 
     public Product registerProduct(Product product) {
+        // '카카오' 문구 검증
+        validateKakao(product.getName());
```
`validateKakao` 라는 메서드의 이름만 보면 어떤 유효성 검사를 수행하는지 쉽사리 짐작이 안되는데요, 조금 더 직관적인 이름을 사용해보는 것은 어떨까요?

지금 네이밍이 너무 이상해서 코멘트를 남기는 것은 아닙니다, 사소하지만 네이밍을 짓는 것에도 고민해보시는 습관을 가져보시면 좋겠어서 코멘트 남겨봅니다. 😄

5seonjae replied:
피드백 감사드립니다! 말씀해주신 대로 validateKakao 라는 이름만으로는

- “어떤 유효성인지?”

- “통과 기준이 무엇인지?”

를 충분히 전달하지 못하네요.
해당 메서드의 실제 역할은 “상품명에 ‘카카오’가 들어가는 경우, 승인 목록(whitelist)에 등록되어 있는지 검사"하는 것입니다.
따라서 다음처럼 행위·대상·조건이 모두 드러나는 이름인 **verifyKakaoNameIsApproved**으로 바꾸겠습니다.

hjjae2 replied:
좋네요! 👍

hjjae2 commented on src/main/resources/schema.sql:
```diff
@@ -4,4 +4,11 @@ CREATE TABLE products
+);
+
+-- "카카오" 문구가 포함된 상품명을 등록할 수 있는 승인 리스트 테이블
+CREATE TABLE approved_products
+(
+    id BIGINT AUTO_INCREMENT PRIMARY KEY,
+    name VARCHAR(100) NOT NULL UNIQUE
```
> 위 방식처럼 별도의 승인 목록 테이블을 만들고, 이름 기반으로 검증하는 방식이 괜찮은 설계인지
> 더 나은 방식이 있다면 조언을 부탁드립니다

다른 방법으로 구현해본다면 어떤 방법이 있을까요? 한번 고민해서 답글을 부탁드려도 될까요?

지금 방법과 고민 후에 나올 새로운 방법의 특징(예: 장, 단점)에 대해 비교해 설명해주시면 좋겠습니다. 🙇

5seonjae replied:
### **현재 방안 - 별도의 승인 목록 테이블을 만들고, 이름 기반으로 검증하는 방식**

장점 : 

1. 데이터 기반 → 배포 없이 INSERT/DELETE 만으로 승인 상품 수정 가능 (MD 전용 화면을 만들기 쉬움)
2. UNIQUE 제약으로 중복 관리가 쉬움
3. Product 테이블과 완전히 분리되어서 서비스 로직(승인 여부) ⇆ 도메인 모델(상품) 간 의존 최소

단점 :

1. 문자열 완전일치 방식 → 띄어쓰기·대소문자·옵션(“카카오프렌즈”, “카카오 프렌즈”) 변화에 취약
2. 이름이 바뀌면 기존 레코드 삭제 + 재승인 필요 → 변경 이력 추적 어려움
3. 승인 기준이 “패턴”이나 “브랜드 ID”로 바뀌면 테이블 구조를 다시 짜야 함

---

현재 방식에 대해 여러 한계점과 단점들이 존재하는 것을 느끼게 되었습니다.
그래서 이를 보완할 수 있는 대안을 곰곰히 생각해 보았습니다.

### **Product 테이블에 brand_requires_approval + approved 플래그 추가**

```sql
ALTER TABLE products
ADD COLUMN approved TINYINT(1) NOT NULL DEFAULT 0,
ADD COLUMN brand_requires_approval TINYINT(1) DEFAULT 0;
```

- brand_requires_approval : 기본값은 0 / 신규 상품 등록 시 승인이 필요한 상품이라면 1, 승인이 불필요한 일반 상품이라면 0
- approved : 기본값은 0 /  MD가 승인 버튼을 누르면 approved ↗

처리 플로우

- 상품 등록
    1) Product 인스턴스 생성
    2) name 문자열에 정책 키워드(예: “카카오”)가 포함되면 brand_requires_approval = 1 로 세팅
    3) approved = 0(기본) 상태로 INSERT

- 승인 대기 조회
    - 백오피스에서 "SELECT * FROM products WHERE brand_requires_approval = 1 AND approved = 0"

- MD 승인
    - MD가 “승인” 클릭 → approved = 1, 필요 시 approved_at, approved_by 도 업데이트

- 프런트/API 노출 
    - 두 플래그가 모두 0 이거나 1인 상품들만 노출

장점 : 

1. 승인 상태·이력를 상품 레코드 안에서 바로 조회 → JOIN 불필요
2. “승인 대기” 목록, 통계 쿼리 작성이 간단
3. Swagger/프론트엔드에 “승인 여부” 필드 노출해 UX 개선 가능

단점 : 

1. “카카오”처럼 문구 규칙이 아닌 상품 단위 규칙으로 해석되어야 의미가 있음
2. 브랜드 정책이 자주 바뀌면 모든 상품 레코드를 UPDATE 해야 함
3. DB 스키마가 상품 도메인과 정책 도메인을 혼용 → 관심사 분리 약화

이외에도 여러 다른 방법들이 있을 것 같은데, 지금 상황에서 떠오르는 대안은 이정도인 것 같습니다.

---

지금과 같은 소규모 환경에서는 별도 승인 목록 테이블 방식을 유지하되, 아래와 같이 운용하면 좋을 것 같습니다.

- 운영 편의성 유지

    - MD 전용 화면이나 간단한 SQL 스크립트로 승인 리스트를 관리

    - 승인 기준이 바뀔 때마다 테이블에 레코드를 추가·삭제

- 모니터링과 로그 남기기

    - 승인·미승인 처리 시 approved_at과 approved_by를 함께 기록하여 변경 이력 확보

    - 승인 실패 사례(띄어쓰기·대소문자 이슈 등)를 주기적으로 점검

- 확장 시점 정의

    - 가입자 수나 상품 등록 빈도가 급격히 늘어날 때

    - 승인 로직이 복잡해지거나 패턴 매칭 요구사항이 생길 때

위 지표에 해당하는 “규모 확장” 시점이 오면, products 테이블에 brand_requires_approval + approved 플래그를 추가하는 방향으로 생각하고 있습니다.

hjjae2 replied:
### ✅ 

좋습니다. 두 가지 방안에 대해 잘 비교해주셨네요. 👍 결정하신 방안으로 가시죠. 저도 합리적이라고 생각합니다.

> _"DB 스키마가 상품 도메인과 정책 도메인을 혼용 → 관심사 분리 약화"_ 이 부분이 걱정이 되신다면 product_id, approved, brand_requires_approval 테이블을 별도로 관리하시면 해소가 될 수 있겠네요. 😅 


### ✅ 

선재님이 두 가지 방안에 대해 소개해수졌고 각각의 장, 단을 비교해 최종 방안을 결정해주셨습니다. 실무라고 가정했다면, 몇 가지 추가 질문을 던져보고 의견을 조금 더 충분히 나눈 뒤 결정했을 거에요. (지금의 상황, 반복입니다.)

선재님에게 말씀드리고 싶었던 것은, 단순히 어떤 방안에 대해 상대방의 조언/의견을 기대하기보다 스스로 어떤 방안이 있고 각 방안의 특징은 무엇이며, 그래서 내 의견은 무엇인지 결정해보는 연습을 해보시면 좋겠기에 질문을 드렸습니다.

참고로, 제가 추가 질문을 던졌다면 아래 정도였을거에요. (질문은 예시이며, 답글을 주실 필요는 없습니다. 😄 )

1. 기존의 방안(이하 방안1)과 추가로 생각해주신 방안(이하 방안2)이 '규모'와 어떤 관계가 있는지 잘 이해가 되지 않습니다. 추가로 설명해주시겠어요?
2. 각 방안의 장,단점을 비교했을 때 방안2가 구현, 확장이나 변경 사항의 대응에 더 유연하다고 판단되는데요, 이해가 다를까요? 
3. _"브랜드 정책이 자주 바뀌면 모든 상품 레코드를 UPDATE 해야 함"_ 정책에 따라 다르다고 생각이 되어 단점의 목록에 포함되어선 안된다고 생각하는데요. 이해가 다를까요?
4. ...

### ✅ 
> 위 방식처럼 별도의 승인 목록 테이블을 만들고, 이름 기반으로 검증하는 방식이 괜찮은 설계인지
> 더 나은 방식이 있다면 조언을 부탁드립니다

저도 선재님이 생각해주신 방안 2를 생각했었어요. 😄

hjjae2 commented on src/test/java/gift/ProductViewControllerTest.java:
```diff
@@ -0,0 +1,139 @@
+                .param("price", "12000")
+                .param("imageUrl", "https://image.com/item.jpg")
+                .contentType(MediaType.APPLICATION_FORM_URLENCODED))
+            .andExpect(status().isOk())
```
승인되지 않은 상품명이 등록될 때, 상태 코드가 `400 BAD Request` 가 아닌 `200 Ok` 가 응답되는 것은 ViewController 이기에 의도하신것이라고 생각하면 될까요-?

5seonjae replied:
네, 의도적으로 200 OK를 사용했습니다.

200 OK 상태코드를 사용한 이유는 다음과 같습니다.

1. ProductViewController는 HTML 폼을 위한 View 계층이라서, 유효성-실패(= MD 미승인) 시에도 요청 자체는 정상입니다.

URL·메서드·폼 필드 등 프로토콜 차원에서는 문제가 없으므로 HTTP 관점에서는 “성공적으로 요청을 수신”했다고 볼 수 있습니다.

2. 서버가 결과 페이지를 반환합니다.

스프링 MVC 기본 흐름(BindingResult.hasErrors())처럼 같은 폼을 다시 렌더링해 오류 메시지를 보여 주는 것이 목적이기 때문에, “작업을 수행하고 결과(수정할 폼 HTML)를 돌려준다”라는 의미로 200 OK가 표준적인 처리라고 생각했습니다.

3. UX 측면에서도 400을 내려 브라우저 기본 에러 페이지를 띄우는 것보다, 같은 화면에서 바로 수정할 수 있게 하는 방식이 일반적입니다.

반면 REST API 엔드포인트(ProductController)에서는 JSON 본문을 소비하는 클라이언트가 있으므로, 동일한 ‘카카오 미승인’ 상황에서 400 Bad Request를 반환하도록 GlobalExceptionHandler에 매핑해 두었습니다.

> 정리

ViewController(폼): 200 OK + 폼 재렌더링으로 사용자에게 즉시 오류 안내

REST API(JSON): 400 Bad Request로 클라이언트(코드)가 오류를 구분

hjjae2 replied:
고민과 이유가 있군요. 훌륭하십니다! 👍

hjjae2 commented on src/test/java/gift/ProductViewControllerTest.java:
`ProductController` 에 대해서도 테스트 코드를 작성해봐주시겠어요? 🙇

5seonjae replied:
넵! @SpringBootTest + @AutoConfigureMockMvc 조합을 그대로 사용해
ProductController 에 대한 통합 테스트도 추가하겠습니다.
같은 전략을 유지하면 설정·러닝커브가 없고, 컨트롤러→서비스→리포지토리→ExceptionHandler까지 엔드투엔드 흐름을 한 번에 검증할 수 있다는 장점이 있어 유지하기로 결정했습니다.

hjjae2 commented on src/test/java/gift/ProductViewControllerTest.java:
```diff
@@ -0,0 +1,139 @@
+@AutoConfigureMockMvc
+@SpringBootTest
+@Transactional
+public class ProductViewControllerTest {
```
>  E2E 테스트 작성 방식에 대해
>  MockMvc를 이용해 HTML 폼 전송 시나리오 기반으로 테스트를 작성했는데,
>  제가 구현한 방식이 E2E 테스트의 목적과 방향에 부합하는지 궁금합니다.

네, 잘 이해하셨고 작성해주셨습니다. 😄 

애플리케이션의 시작지점이라 볼 수 있는 사용자나 프론트엔드 애플리케이션의 입장에서, 
호출될 수 있는 경우(= 시나리오)들에 대해 각각의 응답을 확인하며 테스트 케이스를 잘 만들어주셨어요. 👍

hjjae2 commented on README.md:
PR 본문부터 `README.md` 까지, 기능 요구 사항을 꼼꼼히 잘 작성해주시는군요. 훌륭하십니다. 👍

hjjae2 commented on src/test/java/gift/ProductViewControllerTest.java:
> 테스트 실행 시 나타나는 경고 문구

mockito 를 코드에서 사용하지 않으시더라도 프로젝트에는 포함되어 있습니다. 그레이들 의존성 관계를 살펴보시면 찾으실 수 있을거에요.

<img width="367" alt="image" src="https://github.com/user-attachments/assets/e13e1af4-b5c3-4d0a-9734-bda3b37b1c22"/>


[문서](https://javadoc.io/doc/org.mockito/mockito-core/latest/org.mockito/org/mockito/Mockito.html#0.3)에 나와있는 것 처럼 Java 21 부터 라이브러리가 암묵적으로 javaagent 를 부착하는 것을 제한한다고 하네요. 이를 해소하기 위해선 명시적으로 설정해줘야 하구요.

> _Starting from Java 21, the [JDK restricts the ability of libraries to attach a Java agent to their own JVM](https://openjdk.org/jeps/451)._

지금 동작하는데 이상이 없다면 미션을 진행하시는 과정에서는 WARNING 을 무시하고 진행하셔도 괜찮습니다. 문제가 있거나 여유가 되셔서 WARNING 을 제거해보고 싶으시다면 문서에 나와있는 가이드를 참고해 build.gradle 을 설정해보세요. 👍

hjjae2 reviewed (changes requested):
안녕하세요, 선재님! 이번 단계 꼼꼼하게 잘 진행해주셨네요. 👍 

몇 가지 코멘트 남겼는데요, 확인 부탁드립니다. 🙇

5seonjae committed:
refactor: isApproved() 메서드에서 불필요한 null 검증 제거

5seonjae committed:
refactor: '카카오' 문구 검증 메서드 vaildKakao → verifyKakaoNameIsApproved 이름 변경

5seonjae committed:
test: ProductControllerTest 기본 설정 추가

- @AutoConfigureMockMvc, @SpringBootTest, @Transactional 적용
- MockMvc, ObjectMapper, ProductRepository, ApprovedProductRepository 주입
- ProductController 테스트 작성을 위한 뼈대 클래스 생성

5seonjae committed:
feat: ProductCreateRequestDto에 전체 필드 생성자 추가

- 테스트 코드에서 DTO 객체를 간편하게 생성할 수 있도록 지원
- 기존 유효성 검증 및 매핑 로직에는 영향 없음

5seonjae committed:
fix: ProductRepository.save()에서 생성된 PK 반영

- KeyHolder + jdbcClient.update(keyHolder, "id") 적용해 DB가 생성한 id 조회
- 조회한 id로 새 Product 인스턴스를 만들어 반환하도록 수정
- 이전에는 반환 객체의 id가 null이라 발생하던 NPE 위험 해결

5seonjae committed:
test: ProductControllerTest에 승인된 '카카오' 상품명 등록 성공 케이스 추가

- approved_products에 '카카오 프렌즈 볼펜' 저장 후 201 Created 응답 검증

5seonjae committed:
test: ProductControllerTest에 승인되지 않은 '카카오' 상품 등록 실패 케이스 추가

- '카카오 지갑' 등록 시 400 Bad Request 응답과 오류 메시지("$.error") 검증

5seonjae committed:
test: 상품명 누락 시 등록 실패 테스트 추가

- 빈 상품명("")으로 요청 시 400 Bad Request 및 오류 메시지("$.name") 검증

5seonjae committed:
test: 상품명 15자 초과 등록 실패 케이스 추가

- 16자 이름으로 요청 시 400 Bad Request와 오류 메시지("$.name") 검증

5seonjae committed:
test: 가격 누락 시 상품 등록 실패 케이스 추가

- null 가격으로 요청 시 400 Bad Request와 오류 메시지("$.price") 검증

5seonjae committed:
test: 음수 가격 시 상품 등록 실패 케이스 추가

- -1000원과 같이 음수 가격 요청 시 400 Bad Request 및 오류 메시지("$.price") 검증

5seonjae committed:
test: 이미지 URL 누락 시 상품 등록 실패 케이스 추가

- 빈 imageUrl로 요청 시 400 Bad Request 및 오류 메시지("$.imageUrl") 검증

5seonjae committed:
test: 유효하지 않은 이미지 URL 등록 실패 케이스 추가

- 프로토콜이 없는 URL로 요청 시 400 Bad Request 및 오류 메시지("$.imageUrl") 검증

5seonjae committed:
test: ProductViewControllerTest DisplayName에 [Form] 접두어 추가

- ProductControllerTest([API])와 명확히 구분할 수 있도록 모든 테스트 메서드에 [Form] 접두어 적용
- 테스트 리포트 가독성 및 유형별 필터링 편의성 향상

hjjae2 commented on src/test/java/gift/ProductControllerTest.java:
```diff
@@ -0,0 +1,149 @@
+@AutoConfigureMockMvc
+@SpringBootTest
+@Transactional
+public class ProductControllerTest {
```
테스트 코드 꼼꼼하게 잘 작성해주셨네요! 👍 
다만, 상품 등록 API에 대해서만 테스트 코드가 작성되어있는데요. 나머지 API에 대해서도 작성부탁드릴게요. 🙇

5seonjae replied:
피드백 감사합니다!
급하게 테스트 코드를 작성하느라 등록 기능만 작성을 했네요...
다음 커밋에서 조회(GET), 수정(PUT), 삭제(DELETE) API에 대한 테스트 코드도 추가로 작성하여 반영하겠습니다. 🙇‍♂️

hjjae2 replied:
꼼꼼하게 잘 작성해주셨네요. 👍

hjjae2 commented on src/test/java/gift/ProductControllerTest.java:
```diff
@@ -0,0 +1,149 @@
+        var dto = new ProductCreateRequestDto("카카오 프렌즈 볼펜", 15000, "https://image.com/item.jpg");
+
+        // when & then
+        mockMvc.perform(post("/api/products")
+            .contentType(MediaType.APPLICATION_JSON)
+            .content(objectMapper.writeValueAsString(dto)))
+            .andExpect(status().isCreated());
```
HTTP 상태 코드만 검증하고 있는데요, 응답 본문의 내용은 검증할 필요가 없을까요?

5seonjae replied:
아, 좋은 지적 감사합니다! 🙏

현재는 HTTP 상태 코드만 검증하고 있는데, 말씀하신 대로 응답 본문까지 검증하는 게 API 스펙을 온전히 보장하는 데 더 안전할 것 같습니다.

다음과 같이 jsonPath를 활용해 반환된 JSON의 주요 필드(예 : 생성된 id, name, price, imageUrl) 값을 함께 검증하도록 테스트를 보완하겠습니다.

```java
.andExpect(jsonPath("$.id").isNumber())
.andExpect(jsonPath("$.name").value(dto.getName()))
.andExpect(jsonPath("$.price").value(dto.getPrice()))
.andExpect(jsonPath("$.imageUrl").value(dto.getImageUrl()));
```

ID의 경우, data.sql에 이미 저장된 값이 있거나 테스트 실행 시 DB 상태가 변경될 수 있음을 고려하여, 특정 값을 단언하기보다 숫자 타입 여부만 확인하는 .isNumber()로 검증하도록 할 예정입니다.

hjjae2 reviewed (changes requested):
안녕하세요, 선재님! 코멘트 드린 부분 잘 확인해주셨네요! 👍 

코멘트 추가로 남긴 부분이 있는데요, 이번 단계에서 정리하고 넘어가면 좋겠습니다. 😄 

확인해주시고 다시 리뷰 요청 부탁드릴게요! 🙇

5seonjae committed:
test: ProductControllerTest에 응답 본문 검증 및 id 타입 검증 방식 반영

- createProduct 성공 시 JSON 필드(`id`, `name`, `price`, `imageUrl`) 검증
- `data.sql`의 사전 데이터 및 DB 상태 변경 가능성을 고려하여 `id`는 특정 값 단언 대신 `.isNumber()`로 타입만 확인
- API가 생성된 리소스의 올바른 값을 반환하는지 확인

5seonjae committed:
test: 상품 조회(GET) API 성공 케이스 테스트 추가

5seonjae committed:
test: 상품 조회(GET) API 실패(404) 테스트 추가

5seonjae committed:
feat(GlobalExceptionHandler): PathVariable/RequestParam 타입 불일치 예외 처리 핸들러 추가

- MethodArgumentTypeMismatchException 예외를 전역에서 잡아 400 Bad Request 응답 반환
- 잘못된 파라미터 이름 및 기대 타입 정보를 기반으로 사용자 친화적인 에러 메시지 생성
- 반환 포맷을 Map<String, String>으로 통일하여 `{ "error": "...형식이어야 합니다." }` 형태로 응답

5seonjae committed:
feat(GlobalExceptionHandler): JSON 바디 파싱 오류(HttpMessageNotReadableException) 전역 처리 추가

- HttpMessageNotReadableException을 전역에서 잡아 400 Bad Request 반환
- Jackson의 MismatchedInputException 발생 시
    - 문제 필드명을 `Reference.getFieldName()`으로 추출
    - 기대 타입(`getTargetType().getSimpleName()`)을 함께 사용하여
        `'<field>' 필드는 <Type> 형식이어야 합니다.` 메시지 생성
- 그 외 모든 JSON 파싱 오류는 `"요청 JSON 형식이 잘못되었습니다."` 기본 메시지로 처리
- 응답 포맷을 `Map<String, String>`으로 통일하여 `{ "error": "..." }` 형태 유지

5seonjae committed:
test(ProductController): [API] 잘못된 ID 형식 조회 실패 시나리오 테스트 추가

- 숫자가 아닌 ID("abc")로 조회 요청 시 400 Bad Request 반환 검증
- 응답 JSON의 "error" 필드가 "'id' 파라미터는 long 형식이어야 합니다." 메시지를 포함하는지 확인

5seonjae committed:
docs(GlobalExceptionHandler): JSON 바디 파싱 오류 예외 처리 핸들러 주석 추가

5seonjae committed:
test(ProductController): JSON 파싱 오류(400) 테스트 추가 - 가격 필드에 문자가 들어오는 경우

- 잘못된 JSON 바디(`"price": "과자"`)로 요청했을 때
- GlobalExceptionHandler.handleHttpMessageNotReadable 예외 핸들러가 동작하여
- `"'price' 필드는 Integer 형식이어야 합니다."` 에러 메시지와 400 상태 코드를 반환하는지 검증

5seonjae committed:
test(ProductController): JSON 파싱 오류(400) 테스트 추가 – 값 누락 케이스

- 잘못된 JSON 바디(`"price": ,`)로 요청했을 때
- GlobalExceptionHandler.handleHttpMessageNotReadable가 동작하여
- `"요청 JSON 형식이 잘못되었습니다."` 에러 메시지와 400 상태 코드를 반환하는지 검증

5seonjae committed:
test(ProductController): [API] 상품 수정 성공 테스트 추가

- 기존 상품을 저장한 후 PUT 요청으로 상품 정보를 수정하고
- HTTP 200 OK 응답 및 수정된 id, name, price, imageUrl 필드를 검증

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 없는 ID(404) 테스트 추가

- 존재하지 않는 상품 ID로 PUT 요청 시 404 Not Found 반환 검증
- 응답 JSON의 "error" 필드가 "상품을 찾을 수 없습니다." 메시지를 포함하는지 확인

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 잘못된 ID 형식(400) 테스트 추가

- 숫자가 아닌 ID("abc")로 PUT 요청 시 400 Bad Request 반환 검증
- 응답 JSON의 "error" 필드가 "'id' 파라미터는 long 형식이어야 합니다." 메시지를 포함하는지 확인

5seonjae committed:
refactor(ProductUpdateRequestDto): 생성자 파라미터 price를 int에서 Integer로 변경

- 필드 타입(Integer)와 일치하도록 생성자 시그니처 수정
- null 값 구분 및 Bean Validation 호환성 유지

5seonjae committed:
test(ProductController): 상품 수정 실패 - JSON 파싱 오류(가격 타입 불일치) 테스트 추가

- 잘못된 JSON 바디(`"price": "abc"`)로 PUT 요청 시
- GlobalExceptionHandler.handleHttpMessageNotReadable가 동작하여
- `"'price' 필드는 Integer 형식이어야 합니다."` 메시지와 400 상태 코드를 반환하는지 검증

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 빈 이름 검증 테스트 추가

- 기존 상품 저장 후 빈 문자열 이름으로 PUT 요청 시
- HTTP 400 Bad Request 반환
- 응답 JSON의 "name" 필드에 "상품명은 필수입니다." 메시지 포함 여부 검증

5seonjae committed:
test(ProductController): [API] 상품 등록 실패 - 상품명 허용 문자 검증 테스트 추가

- 상품명에 허용되지 않은 특수문자가 포함된 경우 400 Bad Request 반환 검증
- 응답 JSON의 "name" 필드에 "유효한 특수문자 ( '( )', '[ ]', '+', '-', '&', '/', '_' ) 가 아닙니다." 메시지 포함 여부 확인

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 상품명 15자 초과 테스트 추가

- 16자 이상의 상품명으로 PUT 요청 시 400 Bad Request 반환 검증
- 응답 JSON의 "name" 필드에 "최대 15자까지 가능합니다." 메시지 포함 여부 확인

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 상품명 허용 문자 검증 테스트 추가

- 허용되지 않은 특수문자(‘%’)가 포함된 경우 400 Bad Request 반환 검증
- 응답 JSON의 "name" 필드에 "유효한 특수문자 ( '( )', '[ ]', '+', '-', '&', '/', '_' ) 가 아닙니다." 메시지 포함 여부 확인

5seonjae committed:
test(ProductController): [API] 상품 등록 성공 - 일반 상품명 테스트 추가

- 정상적인 상품명(특수문자 미포함)으로 POST 요청 시
- HTTP 201 Created 응답 검증
- 응답 JSON의 id, name, price, imageUrl 필드가 DTO 값과 일치하는지 확인

5seonjae committed:
test(ProductController): 테스트 메서드명 및 @DisplayName을 “[API] 상품 수정 성공 - 일반 상품명”으로 수정

- 메서드 이름을 `updateProduct_success_normalName`으로 변경
- @DisplayName 어노테이션을 “[API] 상품 수정 성공 - 일반 상품명”으로 업데이트

5seonjae committed:
test(ProductController): [API] 상품 수정 성공 - '카카오' 포함된 승인된 상품명 테스트 추가

- approvedProductRepository에 '카카오 프렌즈 볼펜' 승인 항목 저장
- 해당 이름으로 PUT 요청 시 HTTP 200 OK 및 응답 필드 검증
- id, name, price, imageUrl 필드가 기대값과 일치하는지 확인

5seonjae committed:
chore(ProductControllerTest): 테스트 메서드 순서 재정렬

- 테스트 로직은 그대로 유지

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - '카카오' 포함 승인되지 않은 상품명 테스트 추가

- 승인되지 않은 ‘카카오’ 포함 상품명으로 PUT 요청 시 400 Bad Request 반환 검증
- 응답 JSON의 "error" 필드가 "'카카오'가 포함된 상품은 MD 승인이 필요합니다." 메시지를 포함하는지 확인

5seonjae committed:
test(ProductController): 테스트 메서드명 및 @DisplayName을 "[API] 상품 수정 실패 - 상품명 없음"으로 수정

- 메서드 이름을 `updateProduct_fail_blankName`으로 변경
- @DisplayName 어노테이션을 “[API] 상품 수정 실패 - 상품명 없음”으로 업데이트

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 가격 없음 테스트 추가

- price 필드를 null로 설정하여 PUT 요청 시 400 Bad Request 반환 검증
- 응답 JSON의 "price" 필드에 "가격은 필수입니다." 메시지 포함 여부 확인

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 가격 음수 테스트 추가

- price에 음수 값(-1000)으로 PUT 요청 시 400 Bad Request 반환 검증
- 응답 JSON의 "price" 필드에 "가격은 0 이상이어야 합니다." 메시지 포함 여부 확인

5seonjae committed:
test(ProductController): @DisplayName을 "[API] 상품 수정 실패 - 가격 없음"으로 수정

5seonjae committed:
test(ProductController): @DisplayName을 "[API] 상품 수정 실패 - 가격 음수"으로 수정

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 이미지 URL 없음 테스트 추가

- PUT 요청 시 빈 imageUrl로 400 Bad Request 반환 검증
- 응답 JSON의 "imageUrl" 필드에 "이미지 URL은 필수입니다." 메시지 포함 여부 확인

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - 유효하지 않은 이미지 URL 테스트 추가

- PUT 요청 시 잘못된 형식의 imageUrl("image.com/item.jpg")로 400 Bad Request 반환 검증
- 응답 JSON의 "imageUrl" 필드에 "유효한 이미지 URL이 아닙니다." 메시지 포함 여부 확인

5seonjae committed:
test(ProductController): [API] 상품 수정 실패 - JSON 파싱 오류(400) - 값 누락 테스트 추가

- 잘못된 JSON 바디(imageUrl 값 누락)로 PUT 요청 시 400 Bad Request 반환 검증
- GlobalExceptionHandler.handleHttpMessageNotReadable가 동작하여 "요청 JSON 형식이 잘못되었습니다." 메시지와 400 상태 코드를 반환하는지 확인

5seonjae committed:
test(ProductController): JSON 파싱 오류 테스트에서 잘못된 ID 하드코딩 → 동적 ID 사용으로 수정

- updateProduct_fail_invalidJsonFormat 테스트에 사전 상품 저장 로직 추가
- PUT 요청 시 하드코딩된 `1L` 대신 실제 저장된 `id` 변수를 사용하도록 변경

5seonjae committed:
test(ProductController): [API] 상품 삭제 성공 테스트 추가

- DELETE 요청으로 `/api/products/{id}` 호출 시 204 No Content 반환 검증
- 사전 저장된 상품 ID를 사용하여 삭제 후 상태 코드만 확인

5seonjae committed:
test(ProductController): [API] 상품 삭제 실패 - 없는 ID(404) 테스트 추가

- 존재하지 않는 ID(9999L)로 DELETE 요청 시 404 Not Found 반환 검증

5seonjae committed:
chore(ProductControllerTest): deleteProduct_notFound 테스트에서 unused 변수 제거

- 미사용된 `saved` 변수 삭제하여 코드 간결화

5seonjae committed:
test(ProductController): [API] 상품 삭제 실패 - 잘못된 ID 형식(400) 테스트 추가

- 숫자가 아닌 ID("abc")로 DELETE 요청 시 400 Bad Request 반환 검증

5seonjae committed:
test(ProductController): [API] deleteProduct_notFound 테스트에 에러 메시지 검증 추가

- 불필요한 상품 저장 로직 제거
- 응답 JSON의 "error" 필드에 "상품을 찾을 수 없습니다." 메시지 검증 추가

5seonjae committed:
test(ProductController): deleteProduct_invalidIdFormat 테스트에 에러 메시지 검증 추가

- 불필요한 상품 저장 로직 제거
- 응답 JSON의 "error" 필드가 "'id' 파라미터는 long 형식이어야 합니다."인지 검증하도록 추가

5seonjae committed:
feat(ProductRepository): 테스트용 DB 초기화 지원을 위한 deleteAll() 메서드 추가

- products 테이블의 모든 레코드를 삭제하는 deleteAll() 구현
- 테스트 실행 전 DB 초기화를 위해 사용

5seonjae committed:
chore(ProductControllerTest): 테스트 전 DB 초기화를 위한 @BeforeEach setUp 메서드 추가

- productRepository.deleteAll() 호출로 기존 데이터 제거 설정

5seonjae committed:
test(ProductController): [API] 상품 전체 조회 성공 테스트 추가

- DB 초기화 후 2개 상품 저장
- GET /api/products 호출 시 HTTP 200 OK 반환 및 배열 길이(2)와 각 요소의 name, price 필드 검증

5seonjae committed:
test(ProductViewController): [VIEW] 상품 등록 폼 진입 테스트 추가

- GET /admin/products/new 요청 시
    - HTTP 200 OK 응답
    - 뷰 이름이 "products/form"
    - 모델에 "productRequest" 속성이 존재하며 ProductViewRequestDto 타입인지 검증

5seonjae committed:
chore(ProductControllerTest): @DisplayName “[API] 상품 수정 성공 - 일반 상품명”으로 수정

5seonjae committed:
test(ProductViewController): [Form] 상품 등록 성공 - 일반 상품명 테스트 추가

- POST /admin/products/new 폼 제출 후 302 Found 및 /admin/products 리다이렉트 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 등록 실패 - 허용 문자 검증 테스트 추가

- POST /admin/products/new에서 허용되지 않은 특수문자('%') 사용 시
    - HTTP 200 OK로 form 뷰 유지
    - model에 productRequest의 name 필드 오류 존재 여부 검증

5seonjae committed:
chore(ProductViewControllerTest): 테스트 전 DB 초기화용 productRepository 및 @BeforeEach setUp 추가

- ProductRepository를 주입하여 deleteAll() 호출로 테스트 전 데이터 초기화 설정

5seonjae committed:
test(ProductViewController): [VIEW] 상품 목록 조회 테스트 추가

- GET /admin/products 요청 시
    - 200 OK, 뷰 이름 “products/list”
    - 모델에 “products” 속성 존재 및 size 2 검증
    - 순서대로 name, price, imageUrl 필드 값 검증

5seonjae committed:
test(ProductViewController): [VIEW] 상품 상세 조회 성공 테스트 추가

- GET /admin/products/{id} 요청 시
    - HTTP 200 OK
    - 뷰 이름 "products/detail"
    - 모델에 "product" 속성 존재 여부
    - "product" 객체의 id, name, price, imageUrl 프로퍼티 값 검증

5seonjae committed:
feat(ProductViewController): 상품 상세 조회 실패 시 목록 페이지 리다이렉트 및 플래시 메시지 처리 추가

- viewProductDetail 메서드에 RedirectAttributes 파라미터 추가
- NoSuchElementException 발생 시 "상품을 찾을 수 없습니다." 플래시 메시지 설정 후 /admin/products로 리다이렉트

5seonjae committed:
feat(ProductViewController): 상품 목록 페이지(list.html)에 플래시 에러 메시지 표시 영역 추가

- RedirectAttributes로 전달된 `errorMsg` 플래시 속성이 있을 때 alert 형태로 렌더링

5seonjae committed:
test(ProductViewController): [VIEW] 상품 상세 조회 실패 - 리다이렉트+플래시 테스트 추가

- GET /admin/products/{id} 호출 시 없는 ID에 대해
    - 3xx 리다이렉션 응답
    - /admin/products로 redirect 검증
    - flash 속성 `errorMsg`가 "상품을 찾을 수 없습니다."인지 확인

5seonjae committed:
test(ProductViewController): [VIEW] 상품 상세 조회 실패 테스트에서 리다이렉트 상태코드 검증을 302 Found로 변경

- status().is3xxRedirection()에서 status().isFound()로 수정
- GET /admin/products/{id} 없는 ID 호출 시 302 Found 리다이렉트 검증 유지

5seonjae committed:
test(ProductViewController): [VIEW] 상품 수정 폼 진입 테스트 추가

- GET /admin/products/{id}/edit 요청 시
    - HTTP 200 OK
    - 뷰 이름 "products/form" 반환
    - 모델에 "productRequest" 속성 존재 여부
    - 속성의 name, price, imageUrl 필드 값이 기존 상품과 일치하는지 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 성공 테스트 추가

- POST /admin/products/{id}/edit 유효한 데이터 제출 시
    - 302 Found 응답
    - /admin/products로 리다이렉트 검증

5seonjae committed:
chore(ProductViewControllerTest): updateProduct_success 테스트에서 불필요한 deleteAll() 호출 제거

- 이미 @BeforeEach에서 DB 초기화가 이루어지므로 테스트 메서드 내 deleteAll() 제거

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 성공 - 승인된 '카카오' 상품명 테스트 추가

- approvedProductRepository에 승인된 '카카오 프렌즈 볼펜' 저장
- 해당 이름으로 POST /admin/products/{id}/edit 요청 시
    - 302 Found 응답
    - /admin/products로 리다이렉트 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 실패 - 승인되지 않은 '카카오' 이름 테스트 추가

- 승인되지 않은 '카카오 지갑' 이름으로 POST /admin/products/{id}/edit 요청 시
    - HTTP 200 OK로 폼 뷰 유지
    - model에 `errorMessage` 속성 존재 여부 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 실패 - 상품명 없음 테스트 추가

- POST /admin/products/{id}/edit에서 name 파라미터가 빈 문자열일 때
    - HTTP 200 OK로 products/form 뷰 유지
    - model의 productRequest에 name 필드 에러 존재 여부 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 실패 - 상품명 15자 초과 테스트 추가

- POST /admin/products/{id}/edit에서 16자 이상의 name 파라미터 입력 시
    - HTTP 200 OK로 products/form 뷰 유지
    - model의 productRequest에 name 필드 에러 존재 여부 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 실패 - 상품명에 허용되지 않은 문자 사용 테스트 추가

- POST /admin/products/{id}/edit에서 허용되지 않은 특수문자(‘%’) 포함 name 파라미터 입력 시
    - HTTP 200 OK로 products/form 뷰 유지
    - model의 productRequest에 name 필드 에러 존재 여부 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 실패 - 가격 없음 테스트 추가

- POST /admin/products/{id}/edit에서 price 파라미터가 빈 문자열일 때
    - HTTP 200 OK로 products/form 뷰 유지
    - model의 productRequest에 price 필드 에러 존재 여부 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 실패 - 가격 음수 검증 테스트 추가

- POST /admin/products/{id}/edit에서 price 파라미터가 음수일 때
    - HTTP 200 OK로 products/form 뷰 유지
    - model의 productRequest에 price 필드 에러 존재 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 실패 - 이미지 URL 없음 테스트 추가

- POST /admin/products/{id}/edit에서 imageUrl 파라미터가 빈 문자열일 때
    - HTTP 200 OK로 products/form 뷰 유지
    - model의 productRequest에 imageUrl 필드 에러 존재 여부 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 수정 실패 - 유효하지 않은 이미지 URL 테스트 추가

- POST /admin/products/{id}/edit에서 형식이 잘못된 imageUrl("image.com/item.jpg") 입력 시
    - HTTP 200 OK로 products/form 뷰 유지
    - model의 productRequest에 imageUrl 필드 에러 존재 검증

5seonjae committed:
test(ProductViewController): [Form] 상품 삭제 처리 성공 테스트 추가

- POST /admin/products/{id}/delete 요청 시
    - 302 Found 응답
    - /admin/products로 리다이렉트 검증

5seonjae committed:
docs: README.md에 테스트 및 예외 처리 추가 사항 요약 반영

- ProductViewControllerTest 뷰 레이어 E2E 테스트 항목 정리
- ProductControllerTest REST API 테스트 항목 정리
- GlobalExceptionHandler JSON 파싱·타입 불일치·유효성 오류 핸들러 추가 사항 문서화

5seonjae committed:
docs: README.md에 변경 사항이 아닌 부분 제거 (handleValidationErrors())

5seonjae commented:
가능한 한 모든 경우에 대한 테스트를 추가하려다 보니 시간이 좀 걸렸습니다.
테스트 기능을 구현하면서 등록과 수정 기능에서 중복되는 예외 시나리오가 있었습니다. 
하지만 기능 변경 시 등록과 수정이 각각 독립적으로 영향을 받지 않도록, 같은 예외 케이스라도 테스트를 분리하여 구현했습니다.  

이번 PR에서 다음과 같은 작업을 반영했습니다:

1. View 레이어 E2E 테스트 추가 (`ProductViewControllerTest`)
   - 상품 등록 폼 진입(GET), 등록 처리(POST) 성공·유효성 오류 검증
   - 상품 목록 조회(GET) 정상 검증
   - 상품 상세 조회(GET) 성공·없는 ID 리다이렉트+플래시 검증
   - 상품 수정 폼 진입(GET), 수정 처리(POST) 성공·유효성 오류·‘카카오’ 승인 로직 검증
   - 상품 삭제 처리(POST) 성공 검증

2. REST API 레이어 E2E 테스트 추가 (`ProductControllerTest`)
   - 상품 등록 API(POST): 성공, 유효성 오류(빈값·길이·특수문자·URL), JSON 파싱 오류(타입 불일치·값 누락) 검증
   - 상품 조회 API(GET): 단건 조회(200, 404, 잘못된 ID 형식), 전체 조회(200) 검증
   - 상품 수정 API(PUT): 성공, 404, 잘못된 ID 형식, 유효성 오류, JSON 파싱 오류 검증
   - 상품 삭제 API(DELETE): 성공(204), 404, 잘못된 ID 형식 검증

3. 전역 예외 처리기 추가 (`GlobalExceptionHandler`)
   - `HttpMessageNotReadableException` → JSON 파싱 오류(400 + 메시지)
   - `MethodArgumentTypeMismatchException` → 경로·쿼리 파라미터 타입 불일치(400 + 메시지)

이로써 API 및 뷰 양쪽에서 E2E 수준의 테스트 커버리지를 확보하고, JSON 바디·경로 파라미터 타입 오류 및 유효성 검사 오류를 일관된 방식으로 처리합니다.

5seonjae committed:
docs: README.md에 변경 사항이 아닌 부분 제거

hjjae2 reviewed (approved):
안녕하세요, 선재님! 꼼꼼하게 잘 작성해주셨네요. 고생 많으셨습니다. 👍 

이번 단계는 여기서 머지하고 다음 단계에서 뵙겠습니다. 화이팅입니다! 💪
